factory/src/main/kotlin/io/qalipsis/core/factory/meters/inmemory/StepThroughput.kt

```diff
@@ -0,0 +1,240 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package io.qalipsis.core.factory.meters.inmemory
+
+import com.google.common.util.concurrent.AtomicDouble
+import com.tdunning.math.stats.TDigest
+import io.aerisconsulting.catadioptre.KTestable
+import io.qalipsis.api.lang.supplyIf
+import io.qalipsis.api.meters.DistributionMeasurementMetric
+import io.qalipsis.api.meters.Measurement
+import io.qalipsis.api.meters.MeasurementMetric
+import io.qalipsis.api.meters.Meter
+import io.qalipsis.api.meters.MeterSnapshot
+import io.qalipsis.api.meters.Statistic
+import io.qalipsis.api.meters.Throughput
+import io.qalipsis.api.report.ReportMessageSeverity
+import io.qalipsis.core.factory.meters.MeterSnapshotImpl
+import io.qalipsis.core.reporter.MeterReporter
+import java.time.Duration
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicLong
+import java.util.concurrent.atomic.DoubleAdder
+
+/**
+ * Implementation of meter to track the number of hits measured per a configured unit of time.
+ *
+ * @property timeIncrement the interval of the time slot to calculate the throughput, converted as milliseconds.
+ * @property pastBucketHits the throughput of the lately aggregated time slot
+ * @property intervalCount the count of time slots that were aggregated so far
+ *
+ *
+ * @author Francisca Eze
+ */
+internal class StepThroughput(
+    override val id: Meter.Id,
+    private val meterReporter: MeterReporter,
+    override val unit: ChronoUnit,
+    override val percentiles: Collection<Double>,
+) : Throughput, Meter.ReportingConfiguration<Throughput> {
+
+    private val timeIncrement = Duration.of(1, unit).toMillis()
+
+    @KTestable
+    private var pastBucketHits = AtomicDouble()
+
+    @KTestable
+    internal var tDigestBucket: TDigest? = supplyIf(percentiles.isNotEmpty()) { TDigest.createMergingDigest(100.0) }
+
+    @KTestable
+    private var max = AtomicDouble(0.0)
+
+    internal var intervalCount = 0
+
+    @KTestable
+    private val total = AtomicDouble(0.0)
+
+    @KTestable
+    internal val aggregatedTotal = AtomicDouble(0.0)
+
+    @KTestable
+    private val buckets = List(BUCKET_COUNT) { DoubleAdder() }.toTypedArray()
```

Comment:
More documentation would help to understand how the buckets are working, and what their goals are.